src/pushprocessor.ts

```diff
@@ -162,24 +168,31 @@ const EXPECTED_DEFAULT_UNDERRIDE_RULE_IDS = [
 /**
  * Make sure that each of the rules listed in `defaultRuleIds` is listed in the given set of push rules.
  *
+ * @param kind - the kind of push rule set being merged.
  * @param incomingRules - the existing set of known push rules for the user.
  * @param defaultRules - a lookup table for the default definitions of push rules.
  * @param defaultRuleIds - the IDs of the expected default push rules, in order.
  *
  * @returns A copy of `incomingRules`, with any missing default rules inserted in the right place.
  */
 function mergeRulesWithDefaults(
+    kind: PushRuleKind,
     incomingRules: IPushRule[],
     defaultRules: Record<string, IPushRule>,
-    defaultRuleIds: string[],
+    defaultRuleIds: OrderedRules,
 ): IPushRule[] {
-    // Calculate the index after the last default rule in `incomingRules`
-    // to allow us to split the incomingRules into defaults and custom
-    let firstCustomRuleIndex = incomingRules.findIndex((r) => !r.default);
-    if (firstCustomRuleIndex < 0) firstCustomRuleIndex = incomingRules.length;
-
-    function insertDefaultPushRule(ruleId: string): void {
-        if (ruleId in defaultRules) {
+    // Find the indices of the edges of the user-defined rules in the incoming rules
+    const incomingRulesEnabled = incomingRules.map((rule) => rule.enabled);
+    const userDefinedRulesRange: [number, number] = [
+        incomingRulesEnabled.indexOf(false),
+        incomingRulesEnabled.lastIndexOf(false),
+    ];
```

Comment:
we're actually doing quite a lot of iterating and copying of incomingRules. I'd instead start just by partitioning them:

```
// Split the incomingRules into defaults and custom
const incomingDefaultRules = incomingRules.filter((rule) => rule.default);
const incomingCustomRules = incomingRules.filter((rule) => !rule.default);
```

... and I think everything else can then be more easily written in terms of those, instead of `incomingRules.slice`.

Not a blocker, just a thought.

Reply:
Splitting them as per your example would mean you lose track of the location of the user-defined rules. Right now only `.m.rule.master` precedes them but the server could send other default rules as they are added in that position. This codepath is only hit once for each incoming `m.push_rules` event down /sync, those should be quite rare

Reply:
I don't really follow. We insert the custom rules into `newRules` just before copying `RuleId.SuppressNotices` either way?

> This codepath is only hit once for each incoming m.push_rules event down /sync, those should be quite rare

Yeah, I'm arguing this rather more from the PoV of clarity than performance. Just saying: it's not like doing the `indexOf`s here rather than copying them is saving you much.

Reply:
> I don't really follow. We insert the custom rules into newRules just before copying RuleId.SuppressNotices either way?

incomingRules is in the order of:

```
some default rules (currently just master)
user defined rules
other default rules
```

Your proposed code would merge the default rules into one list and the default orders wouldn't account for any rules in the first bucket other than `master` so the order of any additional rules specified by the server there in later spec versions go into the wrong spot


```
// Split the incomingRules into defaults and custom
const incomingDefaultRules = incomingRules.filter((rule) => rule.default);
const incomingCustomRules = incomingRules.filter((rule) => !rule.default);
```

Reply:
> Your proposed code would merge the default rules into one list and the default orders wouldn't account for any rules in the first bucket other than master so the order of any additional rules specified by the server there in later spec versions go into the wrong spot

I don't think so. In such a scenario, the new rule would end up as the second entry in `incomingDefaultRules` (after `master`, and before `suppress_notices`). Assuming we change the loop below to iterate over `incomingDefaultRules` instead of the current pair of slices of `incomingRules`:
 * We'll see `master`; it will match the expectation, and we will copy it over.
 * We'll see the hypothesized new rule; it is unrecognized, and so we will copy it over.
 * We'll see `suppress_notices`. Its index in orderedRuleIds is higher than the expectation, so first we copy over the custom rules. Then we carry on.

Reply:
Making the proposed change the test fails as follows:

![image](https://github.com/matrix-org/matrix-js-sdk/assets/2403652/07ff9a4b-2089-4303-923e-3e36956d685c)

Reply:
Turns out this was due to missing `".org.matrix.msc3914.rule.room.call"` in the expected defaults in the last PR